Recover from interface assertion panic in convert func

iyabchen opened the pull request:
b/239699320

iyabchen committed:
Recover from interface assertion panic in convert func

melinath commented:
I think you said you were concerned about not being able to test this - but it looks like your tests are working fine? Could you elaborate?

iyabchen commented:
> I think you said you were concerned about not being able to test this - but it looks like your tests are working fine? Could you elaborate?

I am not able to test by deleting a field in .tfplan.json, and trigger the interface assertion. for example, for a  google_compute_instance tfplan, I converted it to .tfplan.json first, then in the file, I delete "description" and its value in .tfplan.json. In `compute_instance.go`, the code was doing `d.Get("description").(string)`, so I thought it should get nil and trigger the error, but it did not. 

The way I can trigger the error is, change `d.Get("description").(string)` -> `d.Get("abc").(string)`, that might match what a terraform provider schema change. 

The unit test I did was, it uses a fake converter with a fake convert function, so it is totally not related to any individual resource converter's logic.

melinath commented:
> The way I can trigger the error is, change `d.Get("description").(string)` -> `d.Get("abc").(string)`, that might match what a terraform provider schema change.

Ah - yeah, I guess that makes sense. If TFV knows the schema type of the field, then it converts nil to the "zero value" for that schema type. (This is the same behavior that TF uses.) https://github.com/GoogleCloudPlatform/terraform-validator/blob/3ef41d069c256419c843e539d2cec36c492cabe7/tfdata/fake_resource_data.go#L77

So the only way to trigger the panic is for the converter to try to access a field that isn't in the schema TFV has - which is what you've done with `d.Get("abc")`.

melinath commented:
That seems odd, though - if that's the only trigger, users should never really run into this, since we keep the TPG / TFV versions in sync. Possibly they're on an older version?

melinath commented:
After thinking about this a little more, I'm not entirely sure that I believe that the TPG and TFV versions being out of sync is the source of this error. It might make sense to just catch panics and display them to users including the resource address so that we can get better debug information from users in the future & more accurately determine a root cause.

iyabchen committed:
error also show resource name

iyabchen committed:
include stack trace with the error

iyabchen committed:
cleanup

iyabchen commented:
> After thinking about this a little more, I'm not entirely sure that I believe that the TPG and TFV versions being out of sync is the source of this error. It might make sense to just catch panics and display them to users including the resource address so that we can get better debug information from users in the future & more accurately determine a root cause.

Updated the code to include stack trace when it recovers from panic, and added the rd.Kind() into the error message.

melinath commented on converters/google/convert.go:
```diff
@@ -233,13 +234,12 @@ func (c *Converter) addDelete(rc *tfjson.ResourceChange) error {
 				continue
 			}
-			return err
+			return fmt.Errorf("failed to convert %s: %s", rd.Kind(), err)
```
in terms of letting the user know the address - I think that once the panic is converted to an error, it will already have the address included here: 

- https://github.com/GoogleCloudPlatform/terraform-validator/blob/53973b8b1e9ed2d9b9d41d5dc747bcf40ea3761b/converters/google/convert.go#L203
- https://github.com/GoogleCloudPlatform/terraform-validator/blob/53973b8b1e9ed2d9b9d41d5dc747bcf40ea3761b/converters/google/convert.go#L210

So it should be safe to leave off the Kind, since that is included in the address, and only return the error. That would keep the error message shorter and easier to understand.

melinath replied:
docs for resource address: https://www.terraform.io/internals/json-format#plan-representation

iyabchen replied:
Done.

melinath commented on converters/google/convert.go:
```diff
@@ -280,20 +280,20 @@ func (c *Converter) addDelete(rc *tfjson.ResourceChange) error {
 				continue
 			}
-			return err
+			return fmt.Errorf("failed to convert %s: %s", rd.Kind(), err)
```
likewise I think just returning the error should be enough

melinath reviewed (changes requested):
A couple minor requests, otherwise LGTM

iyabchen committed:
return a simpler error

iyabchen committed:
fix debug stack not cast to string
